core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -266,8 +266,16 @@ class MetadataCache(brokerId: Int) extends Logging {
 
   def getClusterMetadata(clusterId: String, listenerName: ListenerName): Cluster = {
     val snapshot = metadataSnapshot
-    val nodes = snapshot.aliveNodes.map { case (id, nodes) => (id, nodes.get(listenerName).orNull) }
-    def node(id: Integer): Node = nodes.get(id.toLong).orNull
+    val nodes = snapshot.aliveNodes.flatMap { case (id, nodesByListener) =>
+      nodesByListener.get(listenerName).map { node =>
+        id -> node
+      }
+    }
+
+    def node(id: Integer): Node = {
+      nodes.getOrElse(id.toLong, new Node(id, "", -1))
```

Comment:
How about using a helper method to create ```new Node(id, "", -1)``` ? It can be used by ```MetadataResponse``` also.

Reply:
That's not a bad idea. I think `Node` is public, though we haven't always been strict about requiring KIPs for new methods. A couple options are to add a new constructor or a static factory. We could also add a utility somewhere, but that feels a little clumsy. Thoughts?

Reply:
>  I think Node is public, though we haven't always been strict about requiring KIPs for new methods. A couple options are to add a new constructor or a static factory. We could also add a utility somewhere, but that feels a little clumsy. Thoughts?

You are right.

For another, is this a kind of behavior change to ```ClientQuotaCallback``` (if custom ```ClientQuotaCallback``` does depend on the ```null``` value)? If so, how we keep the compatibility or should we offer a helper method to enable ```ClientQuotaCallback``` to validate the "new Node(id, "", -1)"?

Reply:
I think what we are implementing here is what is documented in `PartitionInfo`:
```java
    /**
     * The complete set of replicas for this partition regardless of whether they are alive or up-to-date
     */
    public Node[] replicas() {
        return replicas;
    }
```
Even the internal logic in `PartitionInfo` assumes non-null values in the array (I caught this issue because of an NPE in the `toString`). We probably can't 100% guarantee that all quota callbacks have appropriate `isEmpty` checks, but my guess is that usage of this plugin is rare in any case and reliance on the replica set is less likely than reliance on the leader. All in all, I think we're better fixing the bug.